### PR TITLE
Ip functions

### PR DESCRIPTION
**Pull request by ciorg**
ip functions
 * isIP
* isIPV6
* isIPV4
* isCIDR
* inIPRange
* isRoutableIP
* isNonRoutableIP
* reverseIP
* isMappedIPV4
* extractMappedIPV4
* IPToInt
* intToIP
* CIDRMin
* CIDRMax
* CIDRBroadcast
* CIDRNetwork
* toCID

There is an issue with the IPtoInt config function tests in data-mate- the test framework returns different values for the `data frame adapter` and the `function adapter` test.  Looks like the bigIntToJSON function subtracts one, I tried to apply that to the `frame adapter`, but since it uses toJSON() the return values are not really big ints.  So that needs to be figured out.  

So that one test won't pass right now.

Other than that, this is ready for review.

### CODE REVIEW

**Comment by peterdemartini** [comment]
I think `IPToInt` should be `ipToInt`

**Comment by peterdemartini** [comment]
I also think this easier to to read:

`CIDRMax` => `getCIDRMax`
`CIDRMin` => `getCIDRMin`
`CIDRBroadcast` => `getCIDRBroadcast`
`CIDRNetwork` => `getCIDRNetwork`

File: packages/data-mate/src/function-configs/ip/CIDRMin.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: '2001:0db8:0123:4567:89ab:cdef:1234:5678/46',
+            output: '2001:db8:120::1',
+        }
+    ],
+    description: 'Returns the first address of a CIDR range, excluding the network address',
+    create() { return CIDRMin; },
+    accepts: [FieldType.String],
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/CIDRMin.ts:38]:
```suggestion
    accepts: [FieldType.String, FieldType.IPRange],
```

This should also define the `output_type` with a field type of `IP`

File: packages/data-mate/src/function-configs/ip/intToIP.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        },
+        {
+            args: { version: '6' },
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: '42540488320432167789079031612388147200',
+            output: '2001:2::',
+        }
+    ],
+    accepts: [FieldType.String],
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/intToIP.ts:29]:
```suggestion
    accepts: [FieldType.String, FieldType.Number],
```

File: packages/data-mate/src/function-configs/ip/isCIDR.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: 'badIPAddress/24',
+            output: null,
+        }
+    ],
+    description: 'Checks if the input is a valid ipv4 or ipv6 ip address in CIDR notation',
+    create() { return isCIDR; },
+    accepts: [FieldType.String],
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/isCIDR.ts:45]:
```suggestion
    accepts: [FieldType.String,  FieldType.IPRange],
```

File: packages/data-mate/src/function-configs/ip/isIP.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: 'not an ip address',
+            output: null,
+        },
+    ],
+    description: 'Checks if the input is a valid ipv4 or ipv6 ip address.  Accepts dot notation for ipv4 addresses and hexadecimal separated by colons for ipv6 addresses',
+    create() { return isIP; },
+    accepts: [FieldType.String]
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/isIP.ts:52]:
```suggestion
    accepts: [FieldType.String, FieldType.IP]
```

File: packages/data-mate/src/function-configs/ip/isIPV4.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: 'not an ip address',
+            output: null,
+        },
+    ],
+    description: 'Checks if the input is a valid ipv4 address in dot notation.',
+    create() { return isIPV4; },
+    accepts: [FieldType.String],
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/isIPV4.ts:45]:
```suggestion
    accepts: [FieldType.String, FieldType.IP],
```

File: packages/data-mate/src/function-configs/ip/isIPV6.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: 'not an ip address',
+            output: null,
+        },
+    ],
+    description: 'Checks if the input is a valid ipv4 or ipv6 ip address.  Accepts dot notation for ipv4 addresses and hexadecimal separated by colons for ipv6 addresses',
+    create() { return isIPV6; },
+    accepts: [FieldType.String],
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/isIPV6.ts:52]:
```suggestion
    accepts: [FieldType.String, FieldType.IP],
```

File: packages/data-mate/src/function-configs/ip/isMappedIPV4.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: 'not an ip address',
+            output: null,
+        },
+    ],
+    description: 'Checks if the input is an ipv4 address mapped to an ipv6 address',
+    create() { return isMappedIPV4; },
+    accepts: [FieldType.String],
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/isMappedIPV4.ts:52]:
```suggestion
    accepts: [FieldType.String, FieldType.IP],
```

File: packages/data-mate/src/function-configs/ip/isNonRoutableIP.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: '2001:2ff::ffff',
+            output: null,
+        },
+    ],
+    description: 'Checks if the input is a non-routable ip address, handles ipv6 and ipv4 address.  Non-routable ip ranges are private, uniqueLocal, loopback, unspecified, carrierGradeNat, linkLocal, reserved, rfc6052, teredo, 6to4, broadcast',
+    create() { return isNonRoutableIP; },
+    accepts: [FieldType.String],
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/isNonRoutableIP.ts:52]:
```suggestion
    accepts: [FieldType.String, FieldType.IP],
```

File: packages/data-mate/src/function-configs/ip/isRoutableIP.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: 'not an ip address',
+            output: null,
+        },
+    ],
+    description: 'Checks if the input is a routable ipv4 or ipv6 address.  Routable ranges are defined as anything that is not in the following ip ranges: private, uniqueLocal, loopback, unspecified, carrierGradeNat, linkLocal, reserved, rfc6052, teredo, 6to4, or broadcast',
+    create() { return isRoutableIP; },
+    accepts: [FieldType.String],
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/isRoutableIP.ts:52]:
```suggestion
    accepts: [FieldType.String, FieldType.IP],
```

File: packages/data-mate/src/function-configs/ip/reverseIP.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: '2001:2::',
+            output: '0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.2.0.0.0.1.0.0.2',
+        }
+    ],
+    description: 'Returns the ip address in reverse notation, accepts both IPv4 and IPv6 addresses',
+    create() { return reverseIP; },
+    accepts: [FieldType.String],
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/reverseIP.ts:38]:
```suggestion
    accepts: [FieldType.String, FieldType.IP],
```

File: packages/data-mate/src/function-configs/ip/toCIDR.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            args: { suffix: '46' },
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: '2001:0db8:0123:4567:89ab:cdef:1234:5678',
+            output: '2001:db8:120::/46',
+        }
+    ],
+    description: 'Returns a CIDR address based on the provided ip and suffix',
+    accepts: [FieldType.String],
```
Comment (peterdemartini) [inline packages/data-mate/src/function-configs/ip/toCIDR.ts:37]:
```suggestion
    accepts: [FieldType.String, FieldType.IPRange],
```

**Review by peterdemartini — CHANGES_REQUESTED** [review]

File: packages/data-mate/src/function-configs/ip/extractMappedIPV4.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: '::122.168.5.18',
+            output: '122.168.5.18',
+        }
+    ],
+    description: 'Returns the ip address in reverse notation, accepts both IPv4 and IPv6 addresses',
+    create() { return extractMappedIPV4; },
+    accepts: [FieldType.String],
```
[the remark this answers is not in the document]
Comment (peterdemartini) [reply packages/data-mate/src/function-configs/ip/extractMappedIPV4.ts]: This still needs "This should also define the output_type with a field type of IP"

File: packages/data-mate/src/function-configs/ip/CIDRMin.ts (hunk 1 above)
Comment (peterdemartini) [reply packages/data-mate/src/function-configs/ip/CIDRMin.ts]: This still needs "This should also define the output_type with a field type of IP"

File: packages/data-mate/src/function-configs/ip/CIDRMax.ts [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: '2001:0db8:0123:4567:89ab:cdef:1234:5678/46',
+            output: '2001:db8:123:ffff:ffff:ffff:ffff:ffff',
+        }
+    ],
+    description: 'Returns the last address of a CIDR range, excluding the broadcast address for IPv4 addresses',
+    create() { return CIDRMax; },
+    accepts: [FieldType.String],
```
[the remark this answers is not in the document]
Comment (peterdemartini) [reply packages/data-mate/src/function-configs/ip/CIDRMax.ts]: This still needs "This should also define the output_type with a field type of IP"

File: packages/data-mate/src/function-configs/ip/CIDRBroadcast.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+            args: {},
+            config: { version: 1, fields: { testField: { type: FieldType.String } } },
+            field: 'testField',
+            input: '1.2.3.4/32',
+            output: '1.2.3.4',
+        }
+    ],
+    description: 'Returns the broadcast address of a CIDR range, only applicable to IPv4 addresses',
+    create() { return CIDRBroadcast; },
+    accepts: [FieldType.String],
```
[the remark this answers is not in the document]
Comment (peterdemartini) [reply packages/data-mate/src/function-configs/ip/CIDRBroadcast.ts]: This still needs "This should also define the output_type with a field type of IP"

**Comment by ciorg** [comment]
Made the requested changes

**Review by peterdemartini — APPROVED** [review]
LGTM

**Pull request merged**